src/validation.cpp

```diff
@@ -4183,6 +4183,18 @@ static bool ContextualCheckBlockHeader(const CBlockHeader& block, BlockValidatio
     if (block.GetBlockTime() <= pindexPrev->GetMedianTimePast())
         return state.Invalid(BlockValidationResult::BLOCK_INVALID_HEADER, "time-too-old", "block's timestamp is too early");
 
+    // Testnet4 only: Check timestamp against prev for difficulty-adjustment
+    // blocks to prevent timewarp attacks (see https://github.com/bitcoin/bitcoin/pull/15482).
+    if (consensusParams.hashGenesisBlock == uint256S("0x00000000da84f2bafbbc53dee25a72ae507ff4914b867c565be350b0da8bf043")) {
+        // Check timestamp for the first block of each difficulty adjustment
+        // interval, except the genesis block.
+        if (pindexPrev->nHeight % consensusParams.DifficultyAdjustmentInterval() == consensusParams.DifficultyAdjustmentInterval() - 1) {
```

Comment:
Is there a reason this can't be simplified to:

```suggestion
        if (nHeight % consensusParams.DifficultyAdjustmentInterval() == 0) {
```

Reply:
This is copied from `GetNextWorkRequired` which doesn't have the `nHeight` helper. The simplification makes sense, or we can add a helper function: https://github.com/bitcoin/bitcoin/pull/29775#discussion_r1654288556

Reply:
I used the simplification here, thanks!